vdirsyncer/sync.py

```diff
@@ -99,107 +102,328 @@ def to_ident_conflict(self, storage):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def get_by_href_b(self, href, default=(None, None)):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def rollback(self, ident):
+        raise NotImplementedError()
+
+
+class _SqliteStatus(_StatusBase):
+    SCHEMA_VERSION = 1
+
+    def __init__(self, path):
+        self._path = path
+        self._conn = sqlite3.connect(path)
+        self._conn.row_factory = sqlite3.Row
+        self._c = None
+
+        if self._is_latest_version():
+            return
+
+        self._conn.execute('DROP TABLE IF EXISTS status')
+        self._conn.execute('DROP TABLE IF EXISTS new_status')
+        self._conn.execute('DROP TABLE IF EXISTS meta')
+
+        self._conn.execute('''CREATE TABLE meta (
+            "version" INTEGER PRIMARY KEY
+        ); ''')
+        self._conn.execute('INSERT INTO meta (version) VALUES (?)',
+                           (self.SCHEMA_VERSION,))
+
+        self._conn.execute('''CREATE TABLE status (
+            "ident" TEXT PRIMARY KEY NOT NULL,
+            "href_a" TEXT NOT NULL,
+            "href_b" TEXT NOT NULL,
+            "hash_a" TEXT NOT NULL,
+            "hash_b" TEXT NOT NULL,
+            "etag_a" TEXT NOT NULL,
+            "etag_b" TEXT NOT NULL
+        ); ''')
+        self._conn.execute('CREATE UNIQUE INDEX by_href_a ON status(href_a)')
+        self._conn.execute('CREATE UNIQUE INDEX by_href_b ON status(href_b)')
+
+        # We cannot add NOT NULL here because data is first fetched for the
+        # storage a, then storage b. Inbetween the `_b`-columns are filled with
+        # NULL.
+        self._conn.execute('''CREATE TABLE new_status (
+            "ident" TEXT PRIMARY KEY NOT NULL,
+            "href_a" TEXT,
+            "href_b" TEXT,
+            "hash_a" TEXT,
+            "hash_b" TEXT,
+            "etag_a" TEXT,
+            "etag_b" TEXT
+        ); ''')
+
+    def _is_latest_version(self):
+        try:
+            return bool(self._conn.execute(
+                'SELECT version FROM meta WHERE version = ?',
+                (self.SCHEMA_VERSION,)
+            ).fetchone())
+        except sqlite3.OperationalError:
+            return False
+
+    @contextlib.contextmanager
+    def transaction(self):
+        assert self._c is None
+        self._c = self._conn
+        yield
+        self._c.execute('DELETE FROM status')
+        self._c.execute('INSERT INTO status '
+                        'SELECT * FROM new_status')
 
     def insert_ident_a(self, ident, props):
-        props_a, props_b = self._new_ident_to_props.get(ident, (None, None))
-        if props_a is not None:
-            raise _IdentAlreadyExists(old_href=props.href,
-                                      new_href=props_a['href'])
-        self._new_ident_to_props[ident] = props.to_status(), props_b
+        # FIXME: Super inefficient
+        old_props = self.get_new_a(ident)
+        if old_props is not None:
```

Comment:
I'd really like this to stay explicit since types have been very confusing in this module in the past.